launcher/modplatform/modrinth/ModrinthPackExportTask.cpp

```diff
@@ -0,0 +1,309 @@
+                QUrl& url = mod->metadata()->url;
+                // most likely some of these may be from curseforge
+                if (!url.isEmpty() && BuildConfig.MODRINTH_MRPACK_HOSTS.contains(url.host())) {
+                    qDebug() << "Resolving" << relative << "from index";
+
+                    QCryptographicHash sha1(QCryptographicHash::Algorithm::Sha1);
+                    sha1.addData(data);
+
+                    ResolvedFile file{ sha1.result().toHex(), sha512.result().toHex(), url.toString(), openFile.size() };
+                    resolvedFiles[relative] = file;
+
+                    // nice! we've managed to resolve based on local metadata!
+                    // no need to enqueue it
+                    continue;
+                }
+            }
+        }
+
+        qDebug() << "Enqueueing" << relative << "for Modrinth query";
+        pendingHashes[relative] = sha512.result().toHex();
+    }
+
+    makeApiRequest();
+}
+
+void ModrinthPackExportTask::makeApiRequest()
+{
+    if (pendingHashes.isEmpty())
+        buildZip();
+    else {
+        QByteArray* response = new QByteArray;
+        task = api.currentVersions(pendingHashes.values(), "sha512", response);
+        connect(task.get(), &NetJob::succeeded, [this, response]() { parseApiResponse(response); });
+        connect(task.get(), &NetJob::failed, this, &ModrinthPackExportTask::emitFailed);
+        task->start();
+    }
+}
+
+void ModrinthPackExportTask::parseApiResponse(const QByteArray* response)
+{
+    // for some reason the task still is successful even if nothing is done
+    if (task == nullptr)
+        return;
+
+    task = nullptr;
+
+    try {
+        const QJsonDocument doc = Json::requireDocument(*response);
+
+        QMapIterator<QString, QString> iterator(pendingHashes);
+        while (iterator.hasNext()) {
+            iterator.next();
+
+            const QJsonObject obj = doc[iterator.value()].toObject();
+            if (obj.isEmpty())
+                continue;
+
+            const QJsonArray files = obj["files"].toArray();
+            if (auto fileIter = std::find_if(files.begin(), files.end(),
+                                             [&iterator](const QJsonValue& file) { return file["hashes"]["sha512"] == iterator.value(); });
+                fileIter != files.end()) {
+                // map the file to the url
+                resolvedFiles[iterator.key()] =
+                    ResolvedFile{ fileIter->toObject()["hashes"].toObject()["sha1"].toString(), iterator.value(),
+                                  fileIter->toObject()["url"].toString(), fileIter->toObject()["size"].toInt() };
+            }
+        }
+    } catch (const Json::JsonException& e) {
+        qWarning() << "Failed to parse versions response" << e.what();
+    }
+    pendingHashes.clear();
+    buildZip();
+}
+
+void ModrinthPackExportTask::buildZip()
+{
+    static_cast<void>(QtConcurrent::run(QThreadPool::globalInstance(), [this]() {
```

Comment:
maybe this function could be written in as a separate function altogether, that instead of communicating with the main thread via queued connections, would just have a return type `std::optional<QString>` corresponding to an error message, or none when the task finished successfully or aborted. Then you could call it with `QtConcurrent::run`, and get a `QFuture` that you connect to a `QFutureWatcher` to react to it when it finishes executing.

There's an example at [`InstanceImportTask`](https://github.com/PrismLauncher/PrismLauncher/blob/1b053032506afc3b7ad9734aea461dc39101ba64/launcher/InstanceImportTask.cpp#L191).

Reply:
How would the progress be set?

Reply:
_i swear i never received a notification about your message, sorry_

[QFutureWatcher](https://doc.qt.io/qt-5/qfuturewatcher.html) has a couple of progress-related methods you can use, including [a signal](https://doc.qt.io/qt-5/qfuturewatcher.html#progressValueChanged) you can connect to and update the overall progress  from the main thread.

Reply:
Still need a way to differentiate between abort and an error.

Reply:
That's the idea for using the `std::optional<QString>` return type though?

Reply:
*differentiate between abort and success

Reply:
You can look at whether `pendingAbort` is set to true (in which case it got aborted), or false, when it was successful.

Reply:
Thanks a lot for helping me works things out! i've kind of managed to do this - but i can't set progress :'(

Reply:
`QtConcurrent::run` explicitly doesn't support setting progress in qt5. So you done what is possible here.